Readme

Driver compilation and installation
===================================

To get the make options type make help
  make help

tn40xx driver Makefile. Available targets:
  all              - build the driver
  clean            - Clean the driver
  help             - Print this help message
  install          - Install driver to system directory
                     usually, it is /lib/modules/VER/kernel/drivers/net
  default          - all

Special notes for Marvell PHYs:
===============================
	This driver package does not include Marvell PHY firmware files. If you own a TN9210 or TN9710 NIC based on Marvell PHY, you will need to obtain the firmware file in .hdr format prior to compiling this driver

	For TN9210 (DeviceID 4024) please extract the file 88X3140-FW-R02-06-03.hdr from 88X3140-FW-R02-06-03.zip
	For TN9710P (DeviceID 4027) please obtain the file x3310fw_0_3_4_0_9445.hdr
	For TN9710Q (DeviceID 4527) please obtain the file e2010fw_0_3_4_0_9445.hdr

	If you are missing these exact files, the driver will compile without support for these PHYs

        ********************************* Important ***************************************
        xxd is required for compiling TN9210 and TN9710 drivers. On some distributions such as Fedora, it is not installed by default and build will fail. Please make sure xxd is installed prior to building the driver
        ***********************************************************************************

General Notes:
==============
      - Supported kernels: 2.6.32 - 5.x

      - make install should be run by a privileged user after running make:
	sudo make install

      - Ethtool support for 5GBASE-T and 2.5GBASE-T options:
	speed 5000 - 5Gbps
	speed 2500 - 2.5Gbps

Distribution Specific Notes:
===========================

      - in SLES11, set 'allow_unsupported_modules 1'  in /etc/modprobe.d/unsupported-modules
      - in SLES12, set 'allow_unsupported_modules 1'  in /etc/modprobe.d/10-unsupported-modules.conf

      - On Ubuntu, if a new ramfs is not rebuilt after make install and previous driver is loaded after reboot:
	sudo update-initramfs -u -k `uname -r`

      - On Fedora27, you may get the following compilation error:
	Makefile:946: *** "Cannot generate ORC metadata for CONFIG_UNWINDER_ORC=y, please install libelf-dev, libelf-devel or elfutils-libelf-devel".  Stop.
	Please follow the recommendation to install the missing package:
	sudo dnf install elfutils-libelf-devel


release_notes
Supported Cards:
================

VID /DID /SVID/SDID	Interface			Print Name
==================================================================
1fc9/4020/1fc9/3015	CX4 				"TN9030 10GbE CX4 Ethernet Adapter"
1fc9/4020/180c/2040	CX4 				"IEI Mustang-200 10GbE Ethernet Adapter"
1fc9/4022/1fc9/3015	SFP+ APM QT2025 		"TN9310 10GbE SFP+ Ethernet Adapter"
1fc9/4022/1186/4d00	SFP+ APM QT2025			"D-Link DXE-810S 10GbE SFP+ Ethernet Adapter"
1fc9/4022/1043/8709	SFP+ APM QT2025			"ASUS XG-C100F 10GbE SFP+ Ethernet Adapter"
1fc9/4022/1432/8103	SFP+ APM QT2025			"Edimax 10 Gigabit Ethernet SFP+ PCI Express Adapter"
1fc9/4024/1fc9/3015	10GBASE-T Pele MV88X3110	"TN9210 10GBase-T Ethernet Adapter"
1fc9/4025/1fc9/3015	NBASE-T Aquantia AQR105		"TN9510 10GBase-T/NBASE-T Ethernet Adapter"
1fc9/4025/1186/2900	10GBASE-T Aquantia AQ2104	"D-Link DXE-810T 10GBase-T Ethernet Adapter"
1fc9/4025/1432/8102	NBASE-T Aquantia AQR105		"Edimax 10 Gigabit Ethernet PCI Express Adapter"
1fc9/4026/1fc9/3015	SFP+ TI TLK10232		"TN9610 10GbE SFP+ Ethernet Adapter"
1fc9/4026/4c52/1000	SFP+ TI TLK10232		"LR-Link LREC6860AF 10 Gigabit Ethernet Adapter"
1fc9/4027/1fc9/3015	NBASE-T Pele MV88X3310	"TN9710P 10GBase-T/NBASE-T Ethernet Adapter"
1fc9/4027/1432/8104	NBASE-T Pele MV88X3310	"Edimax 10 Gigabit Ethernet PCI Express Adapter"
1fc9/4027/1154/0368	NBASE-T Pele MV88X3310	"Buffalo LGY-PCIE-MG Ethernet Adapter"
1fc9/4027/1546/4027	NBASE-T Pele MV88X3310	"IOI GE10-PCIE4XG202P 10Gbase-T/NBASE-T Ethernet Adapter"
1fc9/4027/4c52/1001	NBASE-T Pele MV88X3310	"LR-Link LREC6860BT 10 Gigabit Ethernet Adapter"
1fc9/4027/1baa/3310	NBASE-T Pele MV88X3310	"QNAP LAN-10G1T-THT 10Gbase-T/NBASE-T Ethernet Adapter"
1fc9/4527/1fc9/3015	5GBASE-T Pele MV88E2010	"TN9710Q 5GBase-T/NBASE-T Ethernet Adapter"

- Kernel versions supported: 2.6.32-5.1

Version 0.3.6.17.2 2019/05/12
=============================
- Fix compilation error under kernel 5.1.1

Version 0.3.6.17.1 2019/01/09
=============================
- Fix compilation error under RHEL 8
- Change default installation folder as required by RHEL 8
- Fix kernel 4.20 compilation error (deprecated ethtool_ops.[gs]et_settings)


Version 0.3.6.17 2018/10/10
=============================
- Update 3310P/E2010P to 0.3.4.0
- Update AQR105 to 2.c.72
- Add support for QNAP's TN9710P SVID/SDID and LED definition
- Add support for LR-Link's TN9610 & TN9710P SVID/SDID
- Power Management improvements

Version 0.3.6.16.1 2018/05/24
=============================
- Fixed CentOS compilation errors

Version 0.3.6.16 2018/03/25
===========================
- New memory allocation method for certain applications (Beta version, disabled)
- Update AQR105 firmware to 2.b.e2
- Improvement to CX4 over short trace
- Improved 3310P/E2010P startup routine for 10G
- Improved 3310P/E2010P startup routine for XS708E compatability
- Update 3310P/E2010P to 0.3.3.0
- Add IOI SVID/SDID support
- Add ethtool 2.5G/5G support (but ethtool may need a rebuild)
- Software PTP support (Beta version, disabled)

Version 0.3.6.15 2017/10/18
===========================
- New build method for Marvell PHYs. Please consult the Readme file
- Improved startup procudure for "Phyless" CX4 mode
- Dynamic advertising of MACA for RX flow control
- Support for IEI SVID/SDID added
- Improved memory allocation for memory limited or fragmented applications

Version 0.3.6.14.3 2017/09/24
=============================
- Fix RX Flow Control
- Fix ethtool mtu on Ubuntu 17.04
- Improve single PHY build code
- Fix compilation error for CentOS 7.4

Version 0.3.6.14.2 2017/07/18
=============================
- Fix missing PCP bits in VLAN tag

Version 0.3.6.14.1 2017/06/27
=============================
- Add CX4 card support (TN9030)
- Fix ethtool speed reporting for SFP+ cards
- Fix Hibernate support
- Support Flow Control for TN9710 cards
- Update PHY firmware to 0.2.8.0 for TN9710 cards
- Support Thunderbolt Surprise Remove
- Fix Compilation failed for kernel 4.11

Version 0.3.6.14 2017/03/08
===========================
- Support for Pele 88X3310P and 88E2010P PHYs

Version 0.3.6.13 2016/09/07
===========================
- New adapter naming convention
- Driver debugging options
- Support RHEL6.8
- Support kernel 4.7
- Firmware update for AQR105 and MV88X3110

Version 0.3.6.12 2015/02/19
===========================
- Support TI TLK10232 phy 
- Remove old MDIO interface 
- Support Aquantia AQ2104 phy 
- Merge D-Link special APM firmware and IDs into Jumbo Driver 
- Fix RHEL7 Compilation error 
- Support Speed force (100,1000,2500,5000,10000) 
- Support ethtool -s 
- Support ethtool -p for kernel version < 3.0 & > 3.13 
- Known Issues: 
- - Ehtool Info does not show support for 2.5GBASE-T & 5GBASE-T, but it could be set for supported devices. 

Version 0.3.6.11 2014/4/10
===========================
- Makefile checks is /etc/pm exists 
- Fixed VLAN bug 
- Removed TN8020 PHY support 
- Added WARNING compilation flag. 
- Work around H/W UDP false checksum error 
- Support ethtool -p for kernel version < 3.0 

Version 0.3.6.11 2013/12/8
===========================
- SFP+ 1G modules support 
- Ethtool RX checksum report fix 

Version 0.3.6.10.1 2013/12/25
===========================
- bdx_skb_rx_add_frag() 
- Receive buffer allocation low memory warning 
- Makefile checks is /etc/pm exists 

Version 0.3.6.10 2013/11/4
===========================
- Pele PHY FW 2.4.1.0 
- False H/W tcp checksum error work around 
- Kernel 3.11 support 
- Multiple cards support 
- Updated Ethtool support 

Version 0.3.6.9 2013/09/22
===========================
- Reduced memory foot print by MAC FW size 
- Fixed close crash bug 

Version 0.3.6.8 2013/08/12
===========================
- Initial tn40xx Jumbo Driver release


AQR105_phy.c
#include "tn40.h"
#include "AQR105_phy.h"

#define   mailboxWrite  (0xC000)
#define   phyImageHeaderContentOffset (0x300)
#define   r1 (0x1E)
#define   LINK_LOOP_MAX   			(25)

/*#define AQ_VERBOSE */

void AQR105_register_settings(struct bdx_priv *priv);
int AQR105_mdio_reset(struct bdx_priv *priv, int port, unsigned short phy);
u32 AQR105_link_changed(struct bdx_priv *priv);
void AQR105_leds(struct bdx_priv *priv, enum PHY_LEDS_OP op);

int AQR105_set_speed(struct bdx_priv *priv, signed int speed)
{
	u16 val, port = priv->phy_mdio_port;
	int rVal = 0;

	if (priv->autoneg == AUTONEG_ENABLE) {
		pr_debug("AQR105 speed %d Autoneg\n", speed);
		BDX_MDIO_WRITE(priv, 0x07, 0x10, 0x9101);
		BDX_MDIO_WRITE(priv, 0x07, 0xC400, 0x9C5A);
		BDX_MDIO_WRITE(priv, 0x07, 0x20, 0x1001);
	} else {
		switch (speed) {
		case 10000:	/*10G */
			pr_debug("AQR105 speed 10G\n");
			BDX_MDIO_WRITE(priv, 0x07, 0x10, 0x9001);
			BDX_MDIO_WRITE(priv, 0x07, 0xC400, 0x40);
			BDX_MDIO_WRITE(priv, 0x07, 0x20, 0x1001);
			break;

		case 5000:	/*5G */
			pr_debug("AQR105 speed 5G\n");
			BDX_MDIO_WRITE(priv, 0x07, 0x10, 0x9001);
			BDX_MDIO_WRITE(priv, 0x07, 0xC400, 0x840);
			BDX_MDIO_WRITE(priv, 0x07, 0x20, 0x1);
			break;

		case 2500:	/*2.5G */
			pr_debug("AQR105 speed 10G\n");
			BDX_MDIO_WRITE(priv, 0x07, 0x10, 0x9001);
			BDX_MDIO_WRITE(priv, 0x07, 0xC400, 0x440);
			BDX_MDIO_WRITE(priv, 0x07, 0x20, 0x1);
			break;

		case 1000:	/*1G */
			pr_debug("AQR105 speed 1G\n");
			BDX_MDIO_WRITE(priv, 0x07, 0x10, 0x8001);
			BDX_MDIO_WRITE(priv, 0x07, 0xC400, 0x8040);
			BDX_MDIO_WRITE(priv, 0x07, 0x20, 0x1);
			break;

		case 100:	/*100m */
			pr_debug("AQR105 speed 100m\n");
			BDX_MDIO_WRITE(priv, 0x07, 0x10, 0x101);
			BDX_MDIO_WRITE(priv, 0x07, 0xC400, 0x40);
			BDX_MDIO_WRITE(priv, 0x07, 0x20, 0x1);
			break;

		default:
			pr_err("does not support speed %u\n", speed);
			rVal = -1;
			break;
		}
	}
	if (rVal == 0) {
		/* restart autoneg */
		val = (1 << 12) | (1 << 9) | (1 << 13);
		BDX_MDIO_WRITE(priv, 0x07, 0x0, val);
	}

	return rVal;

}

__init int AQR105_mdio_reset(struct bdx_priv *priv, int port,
			     unsigned short phy)
{
	u16 val, val1, val2;
	int rVal = 0;
	u32 i;
	u8 *image = AQR105_phy_firmware;
	u32 primaryHeaderPtr;
	u32 primaryIramPtr, primaryIramSize, primaryDramPtr, primaryDramSize;
	u32 byteSize, dWordSize, bytePointer;
	u16 msw = 0, lsw = 0, sd, sv;

	BDX_MDIO_WRITE(priv, 0x1E, 0, 1 << 0xF);	/* Soft Reset */
	msleep(10);
	BDX_MDIO_WRITE(priv, 0x1E, 0, 0);
	BDX_MDIO_WRITE(priv, 0x1E, 0xC001, 1 << 0xF);	/* uP Reset */
	msleep(10);
	BDX_MDIO_WRITE(priv, 0x1E, 0xC001, 0);
	msleep(1);

	val1 = bdx_mdio_read(priv, r1, port, 0xC441);

/*Reset start */
	BDX_MDIO_WRITE(priv, 4, 0xC440, 1);	/*Reset XAUI SerDes */
	for (i = 3; i--;) {	/*Wait until XAUI Serdes reset done and clear */
		val = bdx_mdio_read(priv, 4, port, 0xC440);

		if (!val)
			break;
		msleep(1);
	}
	BDX_MDIO_WRITE(priv, 4, 0xC440, 0);	/*clear Reset XAUI SerDes (probably needed) */

	val = bdx_mdio_read(priv, 4, port, 0x0);
	BDX_MDIO_WRITE(priv, 4, 0x0, 0xA040);	/*Reset PHY XS core */
	bdx_mdio_get(priv);	/*Wait until MDIO transaction finished */

	msleep(2);
	BDX_MDIO_WRITE(priv, 4, 0x0, 0x2040);	/*Reset PHY XS core */
	for (i = 10; i--;) {	/*Wait until PHy XS register read correctly */
		val = bdx_mdio_read(priv, 4, port, 0x0);
		if (!(val == 0xFFFF))
			break;
		msleep(1);
	}
/*Reset End */

	BDX_MDIO_WRITE(priv, r1, 0xC452, 1);	/*NVR Daisy Chain Disable */
	BDX_MDIO_WRITE(priv, r1, 0xC441, 0);

	val = 0;
	val |= 1;		/*upRunStall */
	val |= 1 << 6;		/*upRunStallOverride */
	BDX_MDIO_WRITE(priv, r1, 0xC001, val);

  /*--- Read the segment addresses and sizes -----------------*/
	primaryHeaderPtr = (((image[0x9] & 0x0F) << 8) | image[0x8]) << 12;

	primaryIramPtr =
	    (image[primaryHeaderPtr + phyImageHeaderContentOffset + 0x4 + 2] <<
	     16) | (image[primaryHeaderPtr + phyImageHeaderContentOffset + 0x4 +
			  1] << 8) | image[primaryHeaderPtr +
					   phyImageHeaderContentOffset + 0x4];
	primaryIramSize =
	    (image[primaryHeaderPtr + phyImageHeaderContentOffset + 0x7 + 2] <<
	     16) | (image[primaryHeaderPtr + phyImageHeaderContentOffset + 0x7 +
			  1] << 8) | image[primaryHeaderPtr +
					   phyImageHeaderContentOffset + 0x7];
	primaryDramPtr =
	    (image[primaryHeaderPtr + phyImageHeaderContentOffset + 0xA + 2] <<
	     16) | (image[primaryHeaderPtr + phyImageHeaderContentOffset + 0xA +
			  1] << 8) | image[primaryHeaderPtr +
					   phyImageHeaderContentOffset + 0xA];
	primaryDramSize =
	    (image[primaryHeaderPtr + phyImageHeaderContentOffset + 0xD + 2] <<
	     16) | (image[primaryHeaderPtr + phyImageHeaderContentOffset + 0xD +
			  1] << 8) | image[primaryHeaderPtr +
					   phyImageHeaderContentOffset + 0xD];

	if (1 /*AQ_DEVICE_HHD == port->device */ ) {
		primaryIramPtr += primaryHeaderPtr;
		primaryDramPtr += primaryHeaderPtr;
	}
#ifdef AQ_VERBOSE
	pr_info
	    ("Segment Addresses and Sizes as read from the PHY ROM image header:\n\n");
	pr_info("Primary Iram Address: 0x%x\n", primaryIramPtr);
	pr_info("Primary Iram Size: 0x%x\n", primaryIramSize);
	pr_info("Primary Dram Address: 0x%x\n", primaryDramPtr);
	pr_info("Primary Dram Size: 0x%x\n\n", primaryDramSize);
#endif

  /*-- Load IRAM and DRAM ----------------------*/

	/* clear the mailbox CRC */
	BDX_MDIO_WRITE(priv, r1, 0x200, 0x1000);	/*resetUpMailboxCrc */
	BDX_MDIO_WRITE(priv, r1, 0x200, 0x0000);	/*resetUpMailboxCrc */

	/* load the IRAM */
#ifdef AQ_VERBOSE
	pr_info("Loading IRAM:\n");
#endif

/*#define AQ_IRAM_BASE_ADDRESS 0x40000000 */

	BDX_MDIO_WRITE(priv, r1, 0x202, 0x4000);	/*msw */
	BDX_MDIO_WRITE(priv, r1, 0x203, 0x0000);	/*lsw */

	/* set block size so that there are from 0-3 bytes remaining */
	byteSize = primaryIramSize;
	dWordSize = byteSize >> 2;

	bytePointer = primaryIramPtr;
	for (i = 0; i < dWordSize; i++) {
		/* write 4 bytes of data */
		lsw = (image[bytePointer + 1] << 8) | image[bytePointer];
		bytePointer += 2;
		msw = (image[bytePointer + 1] << 8) | image[bytePointer];
		bytePointer += 2;

		BDX_MDIO_WRITE(priv, r1, 0x204, msw);
		BDX_MDIO_WRITE(priv, r1, 0x205, lsw);
		val = bdx_mdio_read(priv, r1, port, 0x200);
		if (val & (1 << 8)) {
			pr_err("Mailbox Busy !!!\n");
			return -1;
		}
		BDX_MDIO_WRITE(priv, r1, 0x200, mailboxWrite);
	}
	/* Note: this final write right-justifies non-dWord data in the final dWord */
	if (byteSize & 0x3) {
		switch (byteSize & 0x3) {
		case 0x1:
			/* write 1 byte of data */
			lsw = image[bytePointer++];
			msw = 0x0000;
			break;

		case 0x2:
			/* write 2 bytes of data */
			lsw =
			    (image[bytePointer + 1] << 8) | image[bytePointer];
			bytePointer += 2;
			msw = 0x0000;
			break;

		case 0x3:
			/* write 3 bytes of data */
			lsw =
			    (image[bytePointer + 1] << 8) | image[bytePointer];
			bytePointer += 2;
			msw = image[bytePointer++];
			break;
		}
		BDX_MDIO_WRITE(priv, r1, 0x204, msw);
		BDX_MDIO_WRITE(priv, r1, 0x205, lsw);
		BDX_MDIO_WRITE(priv, r1, 0x200, mailboxWrite);
	}

	/* load the DRAM */
#ifdef AQ_VERBOSE
	pr_info("Last IRAM data lsw=0x%x msw=0x%x \n", lsw, msw);
	msw = bdx_mdio_read(priv, r1, port, 0x202);
	lsw = bdx_mdio_read(priv, r1, port, 0x203);
	pr_info("Last+1 IRAM address lsw=0x%x msw=0x%x \n", lsw, msw);
	pr_info("Loading DRAM:\n");
#endif

/*#define AQ_DRAM_BASE_ADDRESS 0x3FFE0000 */
	BDX_MDIO_WRITE(priv, r1, 0x202, 0x3FFE);	/*msw */
	BDX_MDIO_WRITE(priv, r1, 0x203, 0x0000);	/*lsw */

	/* set block size so that there are from 0-3 bytes remaining */
	byteSize = primaryDramSize;
	dWordSize = byteSize >> 2;

	bytePointer = primaryDramPtr;
	for (i = 0; i < dWordSize; i++) {
		/* write 4 bytes of data */
		lsw = (image[bytePointer + 1] << 8) | image[bytePointer];
		bytePointer += 2;
		msw = (image[bytePointer + 1] << 8) | image[bytePointer];
		bytePointer += 2;
		BDX_MDIO_WRITE(priv, r1, 0x204, msw);
		BDX_MDIO_WRITE(priv, r1, 0x205, lsw);
		BDX_MDIO_WRITE(priv, r1, 0x200, mailboxWrite);
	}

	/* Note: this final write right-justifies non-dWord data in the final dWord */
	if (byteSize & 0x3) {
		switch (byteSize & 0x3) {
		case 0x1:
			/* write 1 byte of data */
			lsw = image[bytePointer++];
			msw = 0x0000;
			break;

		case 0x2:
			/* write 2 bytes of data */
			lsw =
			    (image[bytePointer + 1] << 8) | image[bytePointer];
			bytePointer += 2;
			msw = 0x0000;
			break;

		case 0x3:
			/* write 3 bytes of data */
			lsw =
			    (image[bytePointer + 1] << 8) | image[bytePointer];
			bytePointer += 2;
			msw = image[bytePointer++];
			break;
		}
		BDX_MDIO_WRITE(priv, r1, 0x204, msw);
		BDX_MDIO_WRITE(priv, r1, 0x205, lsw);
		BDX_MDIO_WRITE(priv, r1, 0x200, mailboxWrite);
	}
#ifdef AQ_VERBOSE
	pr_info("Last DRAM data lsw=0x%x msw=0x%x \n", lsw, msw);
	msw = bdx_mdio_read(priv, r1, port, 0x202);
	lsw = bdx_mdio_read(priv, r1, port, 0x203);
	pr_info("Last+1 DRAM address lsw=0x%x msw=0x%x \n", lsw, msw);
#endif

  /*-- Clear any resets -----------------------------------*/
	BDX_MDIO_WRITE(priv, r1, 0, 0);

  /*-- Release the uP --------------------------------------*/

	val = 1;		/*upRunStall */
	val |= 1 << 6;		/*upRunStallOverride */
	BDX_MDIO_WRITE(priv, r1, 0xC001, val);
	val |= 1 << 15;		/*upReset */
	BDX_MDIO_WRITE(priv, r1, 0xC001, val);

	/* Need to wait at least 100us. */
	msleep(200);
	BDX_MDIO_WRITE(priv, r1, 0xC001, 0x40);
	for (i = 100; i--;) {
		msleep(20);
		val = bdx_mdio_read(priv, r1, port, 0xCC00);
		if (val & (1 << 6))
			break;

	};
	/*if(!(val&(1<<6))){ // Reset timeout
	   pr_err("AQR105 FW reset timeout 1E.CC00=%x\n",val);
	   }; */

	val = bdx_mdio_read(priv, r1, port, 0x20);
	val2 = bdx_mdio_read(priv, r1, port, 0xC885);
	val1 = val >> 8;
	val &= 0xFF;
	val2 &= 0xFF;
	pr_info("AQR105 FW ver: %x.%x.%x\n", val1, val, val2);
/*Enable AQRate speeds */
	msleep(20);
	BDX_MDIO_WRITE(priv, 0x07, 0x10, 0x9101);
	BDX_MDIO_WRITE(priv, 0x07, 0xC400, 0x9C5A);
	BDX_MDIO_WRITE(priv, 0x07, 0x20, 0x1001);
	/* restart autoneg */
	val = (1 << 12) | (1 << 9) | (1 << 13);
	BDX_MDIO_WRITE(priv, 0x07, 0x0, val);

#if 1				/* led setup */
/*
0xC430=f // Act led (right)  black-no link or no act; green-link blink on act
0xC431=80 // Speed led Amber -on on 10G
Tehuti settings
0xC432=C040 // Speed led Green -on on 5G/2.5G/1G (black on 100M)
D-link settings
0xC432=C060 // Speed led Green -on on 5G/2.5/1G/100M
*/
	{
/*ACT Led - off on no link */
		sv = priv->subsystem_vendor;
		sd = priv->subsystem_device;

		if ((sv == 0x105A) && (sd == 0x7203)) {	/* Promise */
			BDX_MDIO_WRITE(priv, r1, 0xC430, 0xc0ef);
			BDX_MDIO_WRITE(priv, r1, 0xC431, 0xC080);
			BDX_MDIO_WRITE(priv, r1, 0xC432, 0xC040);
		} else {
			BDX_MDIO_WRITE(priv, r1, 0xC430, 0x00);
			BDX_MDIO_WRITE(priv, r1, 0xC431, 0x80);
			if ((sv == 0x1186) && (sd == 0x2900)) {	/*isDlink */
				BDX_MDIO_WRITE(priv, r1, 0xC432, 0xC060);
			} else {
				BDX_MDIO_WRITE(priv, r1, 0xC432, 0xC040);
			}
		}
	}
#endif

	return rVal;

}

static int AQR105_get_link_speed(struct bdx_priv *priv)
{
	unsigned short val;	/*, leds; */
	int link = 0;
	u16 ActLnk_MSK = 0;
	u16 sd, sv;

	val = bdx_mdio_read(priv, 7, priv->phy_mdio_port, 0xC810);
	val = bdx_mdio_read(priv, 7, priv->phy_mdio_port, 0xC810);

	val = val >> 9;
	val &= 0x1f;
	if (val == 4) {		/* Link up */
		val = bdx_mdio_read(priv, 7, priv->phy_mdio_port, 0xC800);
		val = bdx_mdio_read(priv, 7, priv->phy_mdio_port, 0xC800);

		val &= 0xF;
		if ((val >> 1) > 0) {
			sv = priv->subsystem_vendor;
			sd = priv->subsystem_device;
			if (!((sv == 0x1186) && (sd == 0x2900))) {	/*isNotDlink */
				ActLnk_MSK = 0xC0E0;
			}
			bdx_mdio_write(priv, 0x1E, priv->phy_mdio_port, 0xC430,
				       0x0f | ActLnk_MSK);
		}
		switch (val >> 1) {
		case 5:	/*SPEED_RES_5GIG */
			link = SPEED_5000;
			pr_debug("AQR105 5G link detected\n");
			break;

		case 4:	/*SPEED_RES_2.5GIG */
			link = SPEED_2500;
			pr_debug("AQR105 2.5G link detected\n");
			break;

		case 3:	/*SPEED_RES_10GIG */
			link = SPEED_10000;
			pr_debug("AQR105 10G link detected\n");
			break;

		case 2:	/* SPEED_RES_1GIG */
			link = SPEED_1000;
			pr_debug("AQR105 1G link detected\n");
			break;

		case 1:	/* SPEED_RES_100M */
			link = SPEED_100;
			pr_debug("AQR105 100M %s link detected\n",
				 (val & 1) ? "" : "Half Duplex");
			break;

		case 0:	/* SPEED_RES_10M */
			link = 0;
			bdx_mdio_write(priv, 0x1E, priv->phy_mdio_port, 0xC430,
				       0x0);
			pr_debug("AQR105 10M %s link detected. Not Supported\n",
				 (val & 1) ? "" : "Half Duplex");
			break;

		}
		priv->errmsg_count = 0;
	} else {
		bdx_mdio_write(priv, 0x1E, priv->phy_mdio_port, 0xC430, 0x0);
		if (++priv->errmsg_count < MAX_ERRMSGS) {
			pr_debug("AQR105 link down.\n");
		}
	}

	return link;

}

u32 AQR105_link_changed(struct bdx_priv *priv)
{
	u32 link, speed;

	speed = AQR105_get_link_speed(priv);
	pr_debug("AQR105_link_changed speed=%u priv->link_speed=%u\n", speed,
		 priv->link_speed);
	if (speed != (u32) priv->link_speed) {
		switch (speed) {
		case SPEED_10000:
			pr_debug("AQR105 10G link detected\n");
			break;
		case SPEED_5000:
			pr_debug("AQR105 5G link detected\n");
			break;
		case SPEED_2500:
			pr_debug("AQR105 2.5G link detected\n");
			break;
		case SPEED_1000:
			pr_debug("AQR105 1G link detected\n");
			break;
		case SPEED_100:
			pr_debug("AQR105 100M link detected\n");
			break;
		case SPEED_10:
			pr_debug("AQR105 10M link detected\n");
			break;
		default:
			pr_debug("AQR105 link down.\n");
			break;
		}
		bdx_speed_changed(priv, speed);
	}
	link = 0;
	if ((!speed) || (!(link = (READ_REG(priv, regMAC_LNK_STAT) & MAC_LINK_STAT)))) {	/* // XAUI link */
		u32 timeout;
		if (speed) {
			timeout = 1000000;	/* 1/5 sec */
			if (priv->link_loop_cnt++ > LINK_LOOP_MAX) {
				bdx_speed_changed(priv, 0);
				priv->link_loop_cnt = 0;
				pr_debug
				    ("AQR105 trying to recover link after %d tries\n",
				     LINK_LOOP_MAX);
			}
		} else {
			timeout = 5000000;	/* 1 sec */
		}
		pr_debug("AQR105 link = 0x%x speed = 0x%x setting %d timer\n",
			 link, speed, timeout);
		WRITE_REG(priv, 0x5150, timeout);
	}

	return link;

}

void AQR105_leds(struct bdx_priv *priv, enum PHY_LEDS_OP op)
{
	u16 dev = 0x1e, led = 0xC430, led_off = 0, led_on = (1 << 8);

	switch (op) {
	case PHY_LEDS_SAVE:
		priv->phy_ops.leds[0] =
		    bdx_mdio_read(priv, dev, priv->phy_mdio_port, led);
		bdx_mdio_write(priv, dev, priv->phy_mdio_port, led, led_off);
		break;

	case PHY_LEDS_RESTORE:
		bdx_mdio_write(priv, dev, priv->phy_mdio_port, led,
			       priv->phy_ops.leds[0]);

		break;

	case PHY_LEDS_ON:
		bdx_mdio_write(priv, dev, priv->phy_mdio_port, led, led_on);

		break;

	case PHY_LEDS_OFF:
		bdx_mdio_write(priv, dev, priv->phy_mdio_port, led, led_off);

		break;

	default:
		pr_debug("AQR105_leds() unknown op 0x%x\n", op);
		break;

	}

}

__init enum PHY_TYPE AQR105_register(struct bdx_priv *priv)
{
	priv->isr_mask =
	    IR_RX_FREE_0 | IR_LNKCHG0 | IR_LNKCHG1 | IR_PSE | IR_TMR0 |
	    IR_RX_DESC_0 | IR_TX_FREE_0;
	priv->phy_ops.mdio_reset = AQR105_mdio_reset;
	priv->phy_ops.link_changed = AQR105_link_changed;
	priv->phy_ops.ledset = AQR105_leds;
	priv->phy_ops.mdio_speed = MDIO_SPEED_6MHZ;
	AQR105_register_settings(priv);

	return PHY_TYPE_AQR105;

}
